File: Medium/Shortest Common Supersequence/shortest-common-supersequence.java
//{ Driver Code Starts
//Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;
class GFG {
	public static void main (String[] args) {
	    
	    //taking input using Scanner class
		Scanner sc=new Scanner(System.in);
		
		//taking total testcases
		int t=sc.nextInt();
		
		sc.nextLine();
		while(t-->0)
		{
		   //taking String X and Y
		   String S[]=sc.nextLine().split(" ");
		   String X=S[0];
		   String Y=S[1];
		   
		   //calling function shortestCommonSupersequence()
		   System.out.println(new Solution().shortestCommonSupersequence(X, Y, X.length(), Y.length()));
		}
	}




       
}
// } Driver Code Ends


//User function Template for Java

class Solution
{
    //Function to find length of shortest common supersequence of two strings.
    public static int shortestCommonSupersequence(String X,String Y,int m,int n)
    {
        int t[][]=new int[m+1][n+1];
        char x[]=X.toCharArray(); char y[]=Y.toCharArray();
        for(int i=0;i<m+1;i++){
            for(int j=0;j<n+1;j++){
                if(i==0||j==0){
                    t[i][j]=0;
                }
                if(i>0 && j>0){
                    if(x[i-1]==y[j-1]){
                        t[i][j]=1+t[i-1][j-1];
                    }
                    else{
                        t[i][j]=Math.max(t[i-1][j],t[i][j-1]);
                    }
                }
            }
        }
        int res=m+n-t[m][n];
        return res;
    }
}

File: Difficulty: Medium/Shortest Common Supersequence/shortest-common-supersequence.cpp
//{ Driver Code Starts
//Initial template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template for C++


class Solution
{
    public:
    int t[101][101];
    //Function to find length of shortest common supersequence of two strings.
    int solve(string &x,string &y,int m,int n){
        if(m==0||n==0)return m+n;
        if(t[m][n]!=-1)return t[m][n];
        if(x[m-1]==y[n-1])return t[m][n]=1+solve(x,y,m-1,n-1);
        else{
            int a= 1+solve(x,y,m-1,n);
            int b= 1+solve(x,y,m,n-1);
            return   t[m][n]=min(a,b);
        }
    }
    int shortestCommonSupersequence(string X, string Y, int m, int n)
    {
     memset(t,-1,sizeof(t));
        return solve(X,Y,m,n);
    }
};

//{ Driver Code Starts.

int main()
{   
    
    int t;
    
    //taking total testcases
    cin >> t;
    while(t--){
    string X, Y;
    //taking String X and Y
	cin >> X >> Y;
	
	//calling function shortestCommonSupersequence()
	Solution obj;
	cout << obj.shortestCommonSupersequence(X, Y, X.size(), Y.size())<< endl;
    }
	return 0;
}


// } Driver Code Ends